File: backend/app/schemas/channel.py
from datetime import datetime
from typing import Optional

from pydantic import BaseModel, Field


class ChannelConnectTelegram(BaseModel):
    phone: str = Field(..., pattern=r"^\+?\d{7,15}$")
    via: str = Field(default="auto", pattern=r"^(auto|sms|app)$")


class ChannelConfirmTelegram(BaseModel):
    channel_id: int = Field(..., gt=0)
    code: str = Field(..., min_length=1, max_length=10)
    phone_code_hash: str = Field(..., min_length=1)


class ChannelResendTelegram(BaseModel):
    channel_id: int = Field(..., gt=0)
    phone_code_hash: str = Field(..., min_length=1)


class ChannelPasswordTelegram(BaseModel):
    channel_id: int = Field(..., gt=0)
    password: str = Field(..., min_length=1)


class ChannelConnectEmail(BaseModel):
    email: str = Field(..., max_length=255)
    password: str = Field(..., min_length=1)
    imap_host: str = Field(..., max_length=255)
    smtp_host: str = Field(..., max_length=255)
    imap_port: int = Field(default=993, ge=1, le=65535)
    smtp_port: int = Field(default=465, ge=1, le=65535)


class ChannelReconnectStored(BaseModel):
    channel_id: int = Field(..., gt=0)


class DownloadResponse(BaseModel):
    new_messages: int
    last_polled_at: Optional[datetime] = None


class ChannelResponse(BaseModel):
    id: int
    type: str
    name: str
    is_connected: bool
    last_polled_at: Optional[datetime] = None
    created_at: Optional[datetime] = None
    updated_at: Optional[datetime] = None

    model_config = {"from_attributes": True}


class ChannelSendReply(BaseModel):
    message_id: int = Field(..., gt=0)
    content: str = Field(..., min_length=1)
    client_request_id: Optional[str] = Field(None, max_length=64)


class ChannelSendMessage(BaseModel):
    contact_id: int = Field(..., gt=0)
    channel: str = Field(..., pattern=r"^(telegram|email)$")
    content: str = Field(..., min_length=1)
    client_request_id: Optional[str] = Field(None, max_length=64)


class ChannelSendNew(BaseModel):
    channel: str = Field(..., pattern=r"^(telegram|email)$")
    recipient: str = Field(..., min_length=1, max_length=255)
    content: str = Field(..., min_length=1)
    client_request_id: Optional[str] = Field(None, max_length=64)

File: frontend/src/types/channel.ts
export interface Channel {
  id: number
  type: string
  name: string
  is_connected: boolean
  last_polled_at: string | null
  created_at: string | null
  updated_at: string | null
}

export interface DownloadResponse {
  new_messages: number
  last_polled_at: string | null
}

export interface ChannelHealth {
  channel_id: number
  type: string
  is_connected_db: boolean
  adapter_alive: boolean
  needs_reconnect: boolean
}

export interface TelegramConnectRequest {
  phone: string
  via?: 'auto' | 'sms' | 'app'
}

export interface TelegramConnectResponse {
  status: string
  message: string
  channel_id?: number
  phone_code_hash?: string
  code_type?: 'app' | 'sms' | 'call'
  next_type?: 'sms' | 'call' | 'flashcall' | 'missed_call'
  timeout?: number
}

export interface TelegramConfirmRequest {
  channel_id: number
  code: string
  phone_code_hash: string
}

export interface TelegramConfirmResponse {
  status: string
  message: string
  channel_id?: number
  phone_code_hash?: string
  retry_count?: number
}

export interface TelegramPasswordRequest {
  channel_id: number
  password: string
}

export interface TelegramResendRequest {
  channel_id: number
  phone_code_hash: string
}

export interface TelegramResendResponse {
  status: string
  phone_code_hash: string
  code_type: 'app' | 'sms' | 'call'
  next_type?: 'sms' | 'call' | 'flashcall' | 'missed_call'
  timeout?: number
  message: string
}

export interface EmailConnectRequest {
  email: string
  password: string
  imap_host: string
  smtp_host: string
  imap_port?: number
  smtp_port?: number
}

export interface SendReplyRequest {
  message_id: number
  content: string
}

export interface SendMessageRequest {
  contact_id: number
  channel: string
  content: string
}
